### PR TITLE
Palette VAS fixes - User Lockout

### PR DESCRIPTION
**Pull request by gyulalaszlo**

### CODE REVIEW

File: app/palette/csrf_storage.py [hunk 1]
```diff
@@ -73,39 +73,103 @@ def escape(val):
...
+
+        self.log(logging.DEBUG, "run_sql(): %s", sql)
+
+        if self._connectionFn is None:
+            self.log(logging.INFO, "No SqlAlchemy connection function - cannot use backing store")
+            return None
+
+        connection = self._connectionFn()
+
+        if self._connectionFn is None:
```
Comment (petispaespea) [inline app/palette/csrf_storage.py]:
```suggestion
        if connection is None:
```

File: app/palette/csrf_storage.py [hunk 2]
```diff
@@ -73,39 +73,103 @@ def escape(val):
...
+        for row in self.run_sql(sql_string, *args):
+            value = row[0]
+
+        self.log(logging.DEBUG, "RETURN VALUE = %s", value)
+        return value
+
+
+    def _table_sql(self):
+        """ The table columns for the table defition to be created """
+        raise NotImplementedError("SqlCsrfStorage::_table_sql() not implemented")
```
Comment (petispaespea) [inline app/palette/csrf_storage.py]:
```suggestion
        raise NotImplementedError("SqlBackedComponent::_table_sql() not implemented")
```

File: app/palette/csrf_storage.py [hunk 3]
```diff
@@ -114,43 +178,199 @@ def delete_old_token():
...
-        if self._connectionFn is None:
-            self.log(logging.ERROR, "SqlAlchemy connection function returned None - cannot use backing store")
-            return None
+class UserLockoutStorage(SqlBackedComponent):
+    """ Generic store for User login attempts implemented on top of SqlAlchemy / PostgreSQL """
+
+    def __init__(self,
+                 connectionFn=None,
+                 table_name="palette_user_lockout",
+                 logger=logging.getLogger('sql-csrf-storage'),
```
Comment (petispaespea) [inline app/palette/csrf_storage.py]:
```suggestion
                 logger=logging.getLogger('sql-lockout-storage'),
```

File: app/palette/csrf_storage.py [hunk 4]
```diff
@@ -114,43 +178,199 @@ def delete_old_token():
...
+        # so we return an empty
+        return {}
+
+
+class UserLockoutMiddleware(LoggingComponent):
+    def __init__(self, app,
+                 storage=EmptyUserLockoutStorage(),
+                 logger=logging,
+                 user_name_param='username'):
+        """ Creates a new CSRF middleware.
```
Comment (petispaespea) [inline app/palette/csrf_storage.py]: Params of __init__ do not match docstring

Comment (gyulalaszlo) [reply app/palette/csrf_storage.py]: fixed in e3752c3

**Pull request merged**